Allow voting for essay option in unique and multiple question

taitus opened the pull request:
## References

Based on and partially extracted from #5827 by [werdenktwas-gmbh](https://github.com/werdenktwas-gmbh)
The original PR implemented three features in one branch:
	1.	New "essay" question type (open text answers).
	2.	Open-text answer option for unique/multiple-choice questions.
	3.	Preliminary results view for admins.

This PR only addresses (2)

## Objectives
Allow voting for essay option in public section

## Visual Changes
- Option text option for unique and multiple questions
<img width="1374" height="1313" alt="Option text option for unique and multiple questions" src="https://github.com/user-attachments/assets/107550b6-f8d5-4eb2-90f3-0eb40fc459fe" />


## Notes
Builds on  #6061

taitus committed:
Unify single/multiple option rendering

Reduce duplication and prepare the code for the next commit.

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -48,28 +48,41 @@ def multiple_choice_help_text
+      label_tag("web_vote_option_#{option.id}") do
+        html = input_tag(option) + option.title
+        html += open_text_tag(option) if option.open_text?
+        html
```
According to the W3C validator:

> The label element may contain at most one button, input, meter, output, progress, select, or textarea descendant.

So it looks like we need to move the `open_text_tag` bit outside the `<label>` tag :thinking:.

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -48,28 +48,41 @@ def multiple_choice_help_text
+    end
+
+    def existing_text_for(option)
+      question.answers.find_by(author: current_user, option: option)&.answer.to_s
```
Right now the answer is lost when submitting an invalid form, so we probably have to use `form.object.answers` as we do in `existing_answer` :thinking:.

By the way, is there a difference between `existing_text_for` and `existing_answer`?

javierm reviewed (commented):
This is just a very high-level review. I'm wondering whether it's better to release version 2.4.0 before finishing this pull request, and including this one in version 2.5.0 :thinking:.

javierm commented on app/views/admin/poll/questions/options/_form.html.erb:
```diff
@@ -18,6 +18,15 @@
 
+  <div class="row">
+    <div class="column">
+      <%= f.check_box :open_text, checked: @option.open_text? %>
```
Do we need the `checked` attribute? 🤔


```suggestion
      <%= f.check_box :open_text %>
```

javierm commented on app/assets/stylesheets/admin/polls/questions/options/table.scss:
```diff
@@ -17,6 +18,12 @@
+
+    &.open-text {
+      @include breakpoint(medium) {
+        width: calc(100% * 2 / 12);
```
What's the purpose of this line? 🤔

taitus replied:
It reserves some width for the new column on medium and larger screens. Without it, the "Open text" header wraps onto two lines. The 2/12 value follows the same proportional layout already used by the description column. If you have a better approach in mind, I'd be happy to change it.

javierm commented on app/assets/stylesheets/admin/polls/questions/options/table.scss:
```diff
@@ -6,7 +6,8 @@
 
-    &.description {
+    &.description,
+    &.open-text {
```
`td.open-text` is going to be either "Yes" or "No". Do we need the `overflow-wrap` rule in this case? 🤔

javierm commented on app/components/admin/poll/questions/options/table_component.html.erb:
```diff
@@ -16,6 +17,7 @@
       <tr id="<%= dom_id(option) %>" class="poll_question_option" data-option-id="<%= option.id %>">
         <td><%= option.title %></td>
         <td class="description"><%= wysiwyg(option.description) %></td>
+        <td class="open-text"><%= t("admin.questions.show.answers.open_text.#{option.open_text}") %></td>
```
Would it make sense to use `shared.yes` and `shared.no`? 🤔

javierm commented on app/views/admin/poll/questions/options/_form.html.erb:
```diff
@@ -18,6 +18,15 @@
+  <div class="row">
+    <div class="column">
+      <%= f.check_box :open_text, checked: @option.open_text? %>
+      <span class="help-text">
```
Shall we add an `ID` to this element, and an `aria-describedby` attribute to the checkbox?

javierm commented on config/locales/en/admin.yml:
```diff
@@ -1171,6 +1171,11 @@ en:
+            field: Open text
+            true: "Yes"
+            false: "No"
+            help: "Allows users to write their own answer when selecting this option. Useful for options like 'Other (please specify)'."
```
I think double quotes are more common than single quotes in the public area 🤔.

```suggestion
            help: "Allows users to write their own answer when selecting this option. Useful for options like \"Other (please specify)\"."
```

Same in the Spanish text.

javierm commented on spec/system/admin/poll/questions/options/options_spec.rb:
```diff
@@ -46,6 +46,27 @@
+      visit admin_question_path(question)
+      click_link "Add answer"
+
+      expect(page).to have_unchecked_field "Allow entering text as answer"
```
Not a strong prefence here, but I think we usually use `have_field (...) checked: false` instead of `have_unchecked_field` 🤔.

javierm commented on spec/system/admin/poll/questions/options/options_spec.rb:
```diff
@@ -46,6 +46,27 @@
+      fill_in "Answer", with: "Other (please specify)"
+      check "Allow entering text as answer"
+
+      expect(page).to have_checked_field "Allow entering text as answer"
```
I think we can remove this line 🤔. We've just used `check`, so the checkbox is obviously checked 🤔.

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -48,31 +52,51 @@ def multiple_choice_help_text
+        input_tag(option) + option.title
+      end
+      html += open_text_tag(option) if option.allows_custom_text?
+      html
```
What do you think about something like this?

```suggestion
      label_with_input(option) + open_text_tag(option)
```

And extract methods accordingly.

taitus replied:
Do you mean something like this?

```
def choice_field(option)
  label_with_input(option) + open_text_tag(option)
end

def label_with_input(option)
  label_tag("web_vote_option_#{option.id}") do
    input_tag(option) + option.title
  end
end

def open_text_tag(option)
  return "".html_safe unless option.allows_custom_text?
  text_area_tag(...)
end
```

The only thing I'm not entirely convinced about is that, when reading `choice_field`, it looks as though `open_text_tag` is always rendered, while the conditional is actually hidden inside `open_text_tag` 🤔 .

Another option could be:

```
def choice_field(option)
  safe_join([
    option_label(option),
    open_text_tag(option) if option.allows_custom_text?
  ].compact)
end

def option_label(option)
  label_tag("web_vote_option_#{option.id}") do
    input_tag(option) + option.title
  end
end
```

Do you have a preference? I'm not entirely sure which approach would be better here.

javierm commented on app/assets/stylesheets/polls/form.scss:
```diff
@@ -23,6 +23,7 @@
 
     label {
       @include radio-or-checkbox-and-label-alignment;
+      flex-wrap: wrap;
```
What's this line for? 🤔 This would allow the label text to appear below the radio button or checkbox, which usually makes it harder to understand that they're related.

javierm commented on app/assets/stylesheets/polls/form.scss:
```diff
@@ -31,6 +32,11 @@
     }
+
+    .open-text {
+      margin-left: $line-height * 0.8;
```
Where does the 0.8 part come from? 🤔.

Other than that:

```suggestion
      margin-#{$global-left}: $line-height * 0.8;
```

taitus replied:
It was chosen to align the textarea with the option text rather than with the checkbox. Using the full `$line-height` moves it slightly too far to the right, while `0.8 * $line-height` matches the current checkbox and label spacing more closely.

Can you think of a better way to express this without using 0.8?

Although this does not remove the magic number, perhaps giving the value a descriptive name would make the intention clearer 🤔 :
```
$option-text-indent: $line-height * 0.8;

.open-text {
  display: block;
  margin-#{$global-left}: $option-text-indent;
  width: calc(100% - #{$option-text-indent});
}
```

What do you think?

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -48,31 +52,51 @@ def multiple_choice_help_text
+        "web_vote[#{question.id}][answer][#{option.id}]",
+        existing_text_for(option),
+        id: "web_vote_option_#{option.id}_text",
+        disabled: disabled?,
```
Do we need this attribute? 🤔 We're already disabling the fieldset when `disabled?` is true, right?

javierm commented on app/models/poll/answer.rb:
```diff
@@ -7,10 +7,19 @@ class Poll::Answer < ApplicationRecord
   validates :question, presence: true
   validates :author, presence: true
-  validates :answer, presence: true
+  validates :answer, presence: true, unless: ->(poll_answer) { poll_answer.option&.allows_custom_text? }
```
Not sure I understand this one 🤔. I'd kind of expect the opposite: optional if the option doesn't allow custom text 🤔.

taitus replied:
This is intentional. `allows_custom_text?` means the additional text is allowed, not required. For open-text options we intentionally allow a blank custom answer, while regular options and generic open questions still require answer to be present. There's already a spec covering this behaviour (stores the selected open-text option even when custom text is blank).

javierm replied:
Then maybe we shouldn't require an answer in regular options 🤔.

javierm commented on app/models/poll/answer.rb:
```diff
@@ -7,10 +7,19 @@ class Poll::Answer < ApplicationRecord
   validates :author, presence: true
-  validates :answer, presence: true
+  validates :answer, presence: true, unless: ->(poll_answer) { poll_answer.option&.allows_custom_text? }
+  validates :answer, length: { maximum: ->(*) { Poll::Answer.answer_max_length }}
```
This also applies to generic open answers, right? If so, what do you think about adding this rule in a different commit? 🤔

javierm commented on app/models/poll/question/option.rb:
```diff
@@ -51,6 +51,10 @@ def with_read_more?
+  def allows_custom_text?
+    open_text?
+  end
+
```
Any reasons to have two different names for the same thing? 🤔

taitus replied:
My intention was to make the conditions easier to read. `allows_custom_text?` felt a bit more expressive of the intended behavior than `open_text?`.

But if you don't think it adds any clarity, I don't have any problem simplifying it and using `open_text?` everywhere.

javierm commented on config/locales/en/general.yml:
```diff
@@ -646,6 +646,7 @@ en:
     description:
       multiple: "You can select a maximum of %{maximum} answers."
     read_more_about: "Read more about:"
+    open_text_aria_label: "Specify answer for %{option}"
```
Usually label texts don't start with a verb 🤔.

```suggestion
    open_text_aria_label: "Answer for %{option}"
```

Same in Spanish.

javierm reviewed (commented):
I've left some comments. Let me know what you think!

taitus committed:
Limit poll answer length

Validate answer length server-side and enforce the same limit in
open-ended question text areas.

taitus committed:
Allow submitting custom text for poll options

Enable users to provide custom text when voting for poll options
that allow custom text. This is useful for "Other (please specify)"
type answers where predefined options are not sufficient.

Note: We use text_area instead of text_field to better accommodate
the 1000 character limit, even though it displays as a single row.

taitus committed:
Stop duplicating option titles in poll answers and partial results

Option votes are identified by option_id, so storing the selected option
title in answer duplicates the same information and requires both fields
to remain consistent.

Reserve answer for user-entered text and require it only for open
questions. Remove the validation that compared answer with the option
titles, since those values are no longer related.

Explicitly clear option when answering open-ended questions.

Existing data remains valid, so no data migration is required.

taitus committed:
Validate poll answers and partial results for option questions

Require records for unique and multiple choice questions to reference
an option.

Open-ended answers remain valid without an option.

taitus committed:
Remove unused answer indexes from poll answers and partial results

The indexes on answer are no longer useful:

- For questions with options, we use option_id which is already indexed
  separately. The (question_id, answer) index on poll_answers and the
  answer index on poll_partial_results were used by Option#total_votes
  until 3a9f761476, when vote counting switched to option_id.
- For open-ended questions, searching by the full answer text is rarely
  needed and seems not performant anyway.

Additionally, the poll_answers index was causing issues with PostgreSQL
btree limits when users submitted long answers, preventing them from
being saved.

javierm commented on spec/models/poll/answer_spec.rb:
```diff
@@ -25,18 +25,33 @@
-    it "is not valid without an answer" do
+    it "is not valid without an answer when question is open-ended" do
+      answer.question = create(:poll_question_open)
+      answer.option = nil
```
Do we need this line? 🤔

javierm commented on spec/models/poll/answer_spec.rb:
```diff
@@ -25,18 +25,33 @@
-      answer.answer = nil
+    it "is valid without answer text when an option is present" do
+      expect(answer).to be_valid
+      expect(answer.answer).to be nil
```
I know this works, but maybe we could write something similar to what the test description says 🤔. That way we won't have to check what the factory does in order to understand the test.

```suggestion
      answer.option = answer.question.options.sample
      answer.answer = nil
      
      expect(answer).to be_valid
```

javierm commented on spec/models/poll/partial_result_spec.rb:
```diff
@@ -2,18 +2,35 @@
+
+    it "is valid without answer text when an option is present" do
+      expect(partial_result).to be_valid
+      expect(partial_result.answer).to be nil
```
I know this works, but maybe we could write something similar to what the test description says 🤔. That way we won't have to check what the factory does in order to understand the test.

javierm commented on app/views/admin/poll/questions/options/_form.html.erb:
```diff
@@ -18,6 +18,16 @@
+      </span>
+    </div>
+  </div>
+
```
Shall we take this chance to move this view to a component? Then we can extract `allows_custom_text_help_id` to a method and move the `help_id` part of the system test to a component.

taitus committed:
Add Poll::Question#option_for to find options by title

Tests repeatedly looked up options with find_by(title:). Add a small
helper to keep specs readable.

taitus committed:
Use option ids in officing results form params

The form keyed each count by the option position in the list,
implicitly assuming params would match given_order. Use option
ids so each value refers to the option itself, not its order.

Co-authored-by: Javi Martín <javim@elretirao.net>

javierm commented on app/components/admin/poll/questions/options/table_component.html.erb:
```diff
@@ -4,6 +4,7 @@
     <tr>
       <th><%= t("admin.questions.show.answers.title") %></th>
       <th scope="col" class="description"><%= t("admin.questions.show.answers.description") %></th>
+      <th scope="col" class="open-text"><%= t("admin.questions.show.answers.allows_custom_text.field") %></th>
```
To avoid duplication in the translation files:

```suggestion
      <th scope="col" class="open-text"><%= Poll::Question::Option.human_attribute_name(:allows_custom_text)
      %></th>
```

Then we can remove the translations for `admin.questions.show.answers.allows_custom_text.field`.

javierm commented on spec/models/poll/question/option_spec.rb:
```diff
@@ -139,4 +144,18 @@
+
+  describe "#allows_custom_text" do
+    it "defaults to false" do
+      option = create(:poll_question_option)
```
Here we might be testing the model or the factory 🤔. To make sure we're testing the model:

```suggestion
      option = Poll::Question::Option.new
```

javierm reviewed (commented):
@taitus More cowbell, I mean, comments!

javierm commented on spec/models/poll/answer_spec.rb:
```diff
@@ -75,18 +90,25 @@
-      create(:poll_question_option, title: "Three", question: create(:poll_question, poll: create(:poll)))
+    it "is not valid when answer exceeds maximum length" do
+      question = create(:poll_question_open)
+      answer = build(:poll_answer, question: question, answer: "a" * (Poll::Answer.answer_max_length + 1))
```
In similar tests, we use the number instead of the method 🤔.

```suggestion
      answer = build(:poll_answer, question: question, answer: "a" * 1001)
```

Same in the other test, only with `1000` and `1001`.

javierm commented on db/schema.rb:
```diff
@@ -1017,7 +1017,6 @@
     t.index ["author_id"], name: "index_poll_answers_on_author_id"
     t.index ["option_id", "author_id"], name: "index_poll_answers_on_option_id_and_author_id", unique: true
     t.index ["option_id"], name: "index_poll_answers_on_option_id"
-    t.index ["question_id", "answer"], name: "index_poll_answers_on_question_id_and_answer"
```
Would it make sense to replace this one with an index by both `question_id` and `option_id`? 🤔 Not sure; just wondering.

javierm commented on spec/models/poll/answer_spec.rb:
```diff
@@ -25,18 +25,33 @@
+      expect(answer.answer).to be nil
+    end
+
+    it "is not valid without an option when question accepts options" do
```
For completeness sake, shall we also add a test checking that it's valid when one of the options allows custom text?

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -56,32 +60,56 @@ def multiple_choice_help_text
+      ])
+    end
+
+    def option_label(option)
```
The name `option_label` might be a bit misleading here, since it displays both an input and a label 🤔. Maybe `option_field`? Although then `choice_field` and `option_field` would be too similar...

javierm commented on app/assets/javascripts/polls/form.js:
```diff
@@ -1,6 +1,14 @@
+      var $textarea = $(".open-text[data-selects='" + $input.attr("id") + "']");
+
+      if ($textarea.length) {
+        $textarea.prop("disabled", !$input.prop("checked") || $input.prop("disabled"));
```
Do we need the second part of the condition? 🤔

javierm commented on app/assets/javascripts/polls/form.js:
```diff
@@ -11,16 +19,32 @@
 
-      $(".poll-form .multiple-choice :checkbox").on("change", function() {
-        var fieldset = $(this).closest("fieldset");
+      $(".poll-form .multiple-choice").off("change.pollsform").on("change.pollsform", ":checkbox", function() {
```
We've never used the `.off` + `.on` pattern. Why do we need it here and not anywhere else? 🤔

By the way, why aren't we using the `change` event? 🤔

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -56,32 +60,56 @@ def multiple_choice_help_text
+        "web_vote[#{question.id}][answer][#{option.id}]",
+        existing_text_for(option),
+        id: "web_vote_option_#{option.id}_text",
+        class: "open-text",
```
The text area now relates to database columns called `allows_custom_text` and `answer`. So I guess the `open-text` class is a bit confusing 🤔. How about `custom-answer`?

Same in other places where we use the `open-text` class.

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -56,32 +60,56 @@ def multiple_choice_help_text
+      text_area_tag(
+        "web_vote[#{question.id}][answer][#{option.id}]",
+        existing_text_for(option),
+        id: "web_vote_option_#{option.id}_text",
```
```suggestion
        id: "web_vote_option_#{option.id}_answer",
```

javierm commented on app/assets/javascripts/polls/form.js:
```diff
@@ -1,6 +1,14 @@
+      var $input = $(input);
+      var $textarea = $(".open-text[data-selects='" + $input.attr("id") + "']");
+
+      if ($textarea.length) {
```
Is this condition necessary? 🤔

javierm commented on app/models/poll/question.rb:
```diff
@@ -80,7 +84,11 @@ def find_or_initialize_user_answer(user, option_id: nil, answer_text: nil)
+      end
     else
       answer.answer = answer_text
     end
```
Would this make sense?

```ruby
    answer.option = question_options.find(option_id) if accepts_options?
    answer.answer = answer_text.presence

    answer
```

javierm commented on app/models/poll/web_vote.rb:
```diff
@@ -64,16 +64,18 @@ def given_answers(params)
       else
         Array(question_params[:option_id]).map do |option_id|
-          question.find_or_initialize_user_answer(user, option_id: option_id)
+          answer_text = answer_param&.dig(option_id)&.to_s&.strip
```
`nil.to_s` exists, so we could write:

```suggestion
          answer_text = answer_param&.dig(option_id).to_s.strip
```

Related rubocop rule: Lint/RedundantSafeNavigation

javierm commented on app/components/officing/results/form_component.html.erb:
```diff
@@ -11,10 +11,10 @@
-          <%= label_tag "questions_#{question.id}_#{i}", option.title %>
-          <%= number_field_tag "questions[#{question.id}][#{i}]", answer_result_value(question.id, i), min: 0 %>
+          <%= label_tag "questions_#{question.id}_#{option.id}", option.title %>
+          <%= number_field_tag "questions[#{question.id}][#{option.id}]", answer_result_value(question.id, option.id), min: 0 %>
```
False positive. We aren't enabling the autocomplete linter.

taitus committed:
Move poll option form to a component

The old partial's `|| @question.id` fallback was redundant because CanCan
builds options through the question association, which always sets question_id.

taitus committed:
Allow options to accept custom text

Add a boolean field to poll question options to indicate whether
they should allow users to enter custom text. This prepares the
backend for allowing "Other (please specify)" type answers.

javierm commented on app/assets/javascripts/polls/form.js:
```diff
@@ -11,16 +17,32 @@
 
       $(".poll-form .multiple-choice :checkbox").on("change", function() {
-        var fieldset = $(this).closest("fieldset");
+        App.PollsForm.updateMultipleChoiceStatus($(this).closest("fieldset")[0]);
```
Is the `[0]` necessary? 🤔


```suggestion
        App.PollsForm.updateMultipleChoiceStatus($(this).closest("fieldset"));
```

Same for the radio buttons.

javierm commented on app/assets/javascripts/polls/form.js:
```diff
@@ -1,6 +1,12 @@
+      var $input = $(input);
+      var $textarea = $(".custom-answer[data-selects='" + $input.attr("id") + "']");
+
+      $textarea.prop("disabled", !$input.prop("checked"));
```
Maybe we can simplify this a little bit 🤔.

```suggestion
      var $textarea = $("[data-selects='" + input.id + "']");

      $textarea.prop("disabled", !input.checked);
```

javierm commented on db/migrate/20250818091122_add_allows_custom_text_to_poll_question_answers.rb:
```diff
@@ -0,0 +1,5 @@
+class AddAllowsCustomTextToPollQuestionAnswers < ActiveRecord::Migration[7.1]
+  def change
+    add_column :poll_question_answers, :allows_custom_text, :boolean, default: false
```
For consistency with the `allow_custom_content` attributes in budget headings:

```suggestion
    add_column :poll_question_answers, :allow_custom_text, :boolean, default: false
```

javierm commented on app/assets/stylesheets/polls/form.scss:
```diff
@@ -31,6 +31,12 @@
     }
+
+    .custom-answer {
+      display: block;
```
Is this line necessary? 🤔

javierm commented on app/components/admin/poll/questions/options/form_component.html.erb:
```diff
@@ -0,0 +1,36 @@
+  <%= render "shared/errors", resource: option %>
+
+  <%= f.hidden_field :given_order,
+                     value: option.persisted? ? option.given_order : option.class.last_position(option.question_id) + 1 %>
```
Could you add to the commit message the reason why we can remove the `|| @question.id` part of this line? :pray:

(Or maybe change this line in a previous commit.)

javierm commented on app/components/officing/results/form_component.html.erb:
```diff
@@ -11,10 +11,10 @@
-          <%= label_tag "questions_#{question.id}_#{i}", option.title %>
-          <%= number_field_tag "questions[#{question.id}][#{i}]", answer_result_value(question.id, i), min: 0 %>
+          <%= label_tag "questions_#{question.id}_#{option.id}", option.title %>
+          <%= number_field_tag "questions[#{question.id}][#{option.id}]", answer_result_value(question.id, option.id), min: 0 %>
```
There's only one test covering this change, and it's a bit hard to follow 🤔. What do you think of this patch?

```diff
diff --git a/spec/system/officing/results_spec.rb b/spec/system/officing/results_spec.rb
index 17fe8ca865..a82f8ad0b3 100644
--- a/spec/system/officing/results_spec.rb
+++ b/spec/system/officing/results_spec.rb
@@ -132,18 +132,17 @@ describe "Officing Results", :with_frozen_time do
     expect(page).to have_content(I18n.l(Date.current.to_date, format: :long))
     expect(page).to have_content(booth_name)
 
-    within("#question_#{question_1.id}_0_result") do
-      expect(page).to have_content("5555")
-      expect(page).not_to have_content("7777")
-    end
+    expect(page).to have_table rows: [
+      { "Answer" => "Yes", "Votes" => "5555" },
+      { "Answer" => "No", "Votes" => "200" }
+    ]
+    expect(page).not_to have_content("7777")
 
     expect(partial_result.reload.answer).to be nil
 
     within("#white_results") { expect(page).to have_content("6") }
     within("#null_results")  { expect(page).to have_content("7") }
     within("#total_results") { expect(page).to have_content("8") }
-    within("#question_#{question_1.id}_0_result") { expect(page).to have_content("5555") }
-    within("#question_#{question_1.id}_1_result") { expect(page).to have_content("200") }
     expect(page).not_to have_content "What do you want?"
   end
 end
```

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -56,32 +60,56 @@ def multiple_choice_help_text
+    end
+
+    def answer_for(option)
+      answers_for_question.find { |a| a.option_id == option.id }
```
Why have we renamed the `answer` block parameter to `a`? We usually use more descriptive names 🤔.

```suggestion
      answers_for_question.find { |answer| answer.option_id == option.id }
```

Or, alternatively (haven't tested it):

```suggestion
      answers_for_question.find_by(option_id: option.id)
```

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -40,8 +40,12 @@ def options_read_more_links
     end
 
+    def answers_for_question
+      form.object.answers[question.id] || []
```
In order to return similar objects in both cases:

```suggestion
      form.object.answers[question.id] || question.answers.none
```

taitus replied:
After a failed submit, `WebVote#update` replaces `@answers` with Arrays from `given_answers`.

javierm commented on app/components/polls/questions/question_component.rb:
```diff
@@ -56,32 +60,56 @@ def multiple_choice_help_text
+        class: "custom-answer",
+        maxlength: Poll::Answer.answer_max_length,
+        rows: 1,
+        "aria-label": t("poll_questions.open_text_aria_label", option: option.title),
```
Since we've renamed `open-text` to `custom-answer`, should we also rename the I18n key? 🤔

javierm commented on app/controllers/officing/results_controller.rb:
```diff
@@ -51,7 +51,7 @@ def build_results
           )
           partial_result.officer_assignment_id = @officer_assignment.id
-          partial_result.answer = option.title
+          partial_result.answer = nil
```
Do we need this line? 🤔 I know we've talked about it, but I don't remember our conclusion 😌.

javierm commented on app/models/poll/partial_result.rb:
```diff
@@ -9,11 +9,9 @@ class Poll::PartialResult < ApplicationRecord
   validates :question, presence: true
   validates :author, presence: true
-  validates :answer, presence: true
-  validates :answer, inclusion: { in: ->(partial_result) { partial_result.option.possible_answers }},
```
Does this mean we can remove the `possible_answers` method in `Poll::Question::Option` as well?

javierm reviewed (commented):
@taitus Here are a few more comments. Let me know what you think!

javierm commented on app/assets/javascripts/polls/form.js:
```diff
@@ -11,16 +16,32 @@
+        App.PollsForm.syncOpenText(this);
+      });
     },
+
```
Shall we remove this line? We usually don't add blank lines between object properties 🤔.

javierm commented on spec/components/admin/poll/questions/options/form_component_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    render_inline component
+
+    expect(page).to have_field "Answer"
+    expect(page).to have_css "label", text: "Description"
```
In order to check that the label is correctly associated to its input:

```suggestion
    expect(page).to have_field "Description (optional)"
```

javierm commented on spec/components/admin/poll/questions/options/form_component_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    it "includes accessibility attributes" do
+      render_inline component
+
+      checkbox = page.find_field("Allow custom text")
```
```suggestion
      checkbox = page.find_field("Allow custom text", type: :checkbox)
```

javierm commented on spec/components/polls/questions/question_component_spec.rb:
```diff
@@ -80,11 +84,13 @@
     it "selects the option when users have already voted" do
-      create(:poll_answer, author: user, question: question, option: option_yes)
+      option_yes.update!(allow_custom_text: true)
+      create(:poll_answer, author: user, question: question, option: option_yes, answer: "because yes")
```
"Because yes" sounds a bit strange in English 🤔.

```suggestion
      create(:poll_answer, author: user, question: question, option: option_yes, answer: "because")
```

javierm commented on spec/components/polls/questions/question_component_spec.rb:
```diff
@@ -80,11 +84,13 @@
       render_inline Polls::Questions::QuestionComponent.new(question, form: form)
 
       expect(page).to have_field "Yes", type: :radio, checked: true
+      expect(page).to have_field "Answer for Yes", type: :textarea, with: "because yes"
```
```suggestion
      expect(page).to have_field "Answer for Yes", type: :textarea, with: "because"
```

javierm commented on spec/factories/polls.rb:
```diff
@@ -66,6 +66,13 @@
       end
     end
 
+    trait :yes_no_custom_text do
+      after(:create) do |question|
+        create(:poll_question_option, question: question, title: "Yes", allow_custom_text: true)
+        create(:poll_question_option, question: question, title: "No", allow_custom_text: true)
+      end
+    end
```
Do we need to create the `votation_type_unique` here, like we do in the `:yes_no` trait? 🤔 No idea, and maybe it wasn't necessary in the `:yes_no` trait. It's just that the difference feels a bit strange.

javierm commented on spec/models/poll/partial_result_spec.rb:
```diff
@@ -2,18 +2,37 @@
+
+    it "is valid with an option for the question" do
+      question = create(:poll_question, :yes_no)
+      option = question.question_options.first
```
Is this test redundant? Seems to be similar to the "is valid without answer text when an option is present" test 🤔.

javierm commented on spec/system/polls/polls_spec.rb:
```diff
@@ -200,7 +200,8 @@
       poll.update!(geozone_restricted_to: [geozone])
-      create(:poll_answer, author: user, question: question, answer: "Yes")
+      option_yes = question.option_for("Yes")
+      create(:poll_answer, author: user, question: question, option: option_yes)
```
To make more obvious what's changing:

```suggestion
      create(:poll_answer, author: user, question: question, option: question.option_for("Yes"))
```

javierm commented on spec/system/polls/polls_spec.rb:
```diff
@@ -231,7 +232,8 @@
 
-      create(:poll_answer, author: user, question: question, answer: "Answer A")
+      option_a = question.option_for("Answer A")
+      create(:poll_answer, author: user, question: question, option: option_a)
```
```suggestion
      create(:poll_answer, author: user, question: question, option: question.option_for("Answer A"))
```

javierm commented on spec/system/polls/votation_types_spec.rb:
```diff
@@ -51,9 +51,78 @@
+      expect(page).to have_field "Answer for Yes", type: :textarea, disabled: false
+      expect(page).to have_field "Answer for No", type: :textarea, disabled: true
+
+      fill_in "Answer for Yes", with: "because yes"
```
```suggestion
      fill_in "Answer for Yes", with: "because"
```

javierm commented on spec/system/polls/votation_types_spec.rb:
```diff
@@ -51,9 +51,78 @@
     question = create(:poll_question_multiple, :abc, poll: poll, max_votes: 2)
-    create(:poll_answer, author: author, question: question, answer: "Answer A")
+    option_a = question.option_for("Answer A")
+    create(:poll_answer, author: author, question: question, option: option_a)
```
```suggestion
    create(:poll_answer, author: author, question: question, option: question.option_for("Answer A"))
```
